### ISSUE TITLE
App state migration

### ISSUE THREAD

**Issue by philrz** [brimsec/brim#751]
In #587 we introduced versioning of the app state object. A comment from @jameskerr on follow-on work we'd now like to take care of:

> Depending on how critical it is for users to save state, like their history and in the future, their saved queries. We may want to write more code to migrate data and what not.

Indeed, we've warned users through two upgrades that they're going to lose their History, and no doubt they'd appreciate being able to start relying on this data going forward. So let's make migration a thing.

Another comment from there that might be worth taking care of at the same time, if it isn't already:

> I think it would be good to write a test that does some of the common operations on the redux store, the saves the resulting state in a snapshot. Then each time we make changes to the state shape, hopefully that test will fail with a message like "update the state version".

**Comment by jameskerr** [comment]
# State Migrations

The app can break when the redux state changes its shape between versions. The updated code expects the new state, while the saved session contains the old state.

There are a few ways that state can change: 

* Add a reducer
* Add a property to existing reducer
* Change name of existing property
* Remove property from existing reducer
* Remove a reducer

**Adding a reducer**

The app will not break if we add a new reducer because all reducers are given a default initialState if they are called with undefined as the first argument. In this case, a previous session  will pass “undefined” to the new reducer. The new reducer will use the initial state and all will be well.

**Adding a property to an existing reducer**

This will break the app if the code does not expect the property to be undefined. It will break even if you add a default for the property in the initialState (this has bitten me before). The app boots up with the old session data, it sees that it already has state for that reducer so the initialState is not used. Unless the property is truly optional, it’s cumbersome, error-prone, and confusing to assume it can be undefined everywhere.

To consider all options, we could deep merge the initialState with the passed in state, but that will hurt performance as all the reducers get called for every action that we dispatch. 

**Change name of existing property**

This will break the app as well, unless the app knows that the value could be at either name.

**Remove a property from a reducer**

This will not break the app if it no longer reads from that bit of state.

**Remove a reducer**

This will not break the app if it no longer reads from that bit of state. We will however get a warning from redux in the console saying that it did not expect that key to be in the state.

## A thought on migrating down

At first I thought that we should only support migrating up. This will the common case for users. However, it might be good to have. In our last release, we accidentally introduced a show-stopping bug. If the user upgraded to this new release, encounter the bug, then download the old version. The app state will be on the newer version while the app is at the older version. A might work better. During development, I could also easily switch branches and not have to blow away my state each time. 

Wait...this actually doesn't work because the down migration function would only be defined in the newer versions of the code. In light of that, it might be smart to archive the old state right before migrating in the app data directory with the timestamp of the version that it was on. That way if the user downloaded the older version of the app we can load that archived state instead of the new state.

## A migration workflow

1. I change the name of property “packet” to “pcap” in a reducer
2. I run `bin/gen migration rename_packet_to_pcap`
3. A file is created in the migrations directory with the template `migrations/<timestamp>_<name>.js`. For example `migrations/202005130405_renamePacketToPcap.js`
4. In this folder we export a function that will receive the old state and return the new state.

That’s all the developer needs to do.

When the app starts up, it will look at the migrations directory, get the names of all the files, and sort them. This will be oldest to newest since the name always starts with a timestamp. We’ll require that file to get the migration function and extract the timestamp from file name. These timestamps will be the values of state.version. Then end result is a sorted array of objects that looked like `{migration: (prev) => next, version: string}`. 

Then app will look at the version property of the previous session’s state. It will use it to get all the pending migrations (those with a higher version string). It then runs each of those migrations. The result of each becomes the argument for the next. After each migration, we set the version in the state to the version of the migration file. 

If there are pending migrations, let’s make a copy of the old session with a file name that includes its version. This way if a migration fails, we can see what the old state was. It will also be helpful if we downgrade. The app will see that it’s session version is higher that latest migration and know that it has downgraded. It will take the last migration’s version and see if there is an archived session with that version available to load. Otherwise, start up with no previous state.

## Testing

A crude way to test that a migration is needed is to dispatch some common actions, then compare the state to an expected snapshot. Each time we change the shape that snapshot compare will fail, reminding us to write a migration. Otherwise, it's on us as developers to remember the situations in which we need to migrate the data.

Testing migrations will be more difficult. One way to do it would be to checkout the app at an earlier state, download a copy of the state and save it to a json file. To do this we could use the same mechanism that saves an archive of the session described above. We can provide this previous state to the migration function in a test and assert that what we expected takes place. This might be smart for complex migrations. Once verified, the tests can be deleted.

Before each release, it will be wise to checkout the previous release, perform some actions to change the state, quit the app, checkout the new version, and verify all looks well.

## Handling errors

If migrating throws an error we should throw the error but provide a way to delete the previous session as an escape hatch. If the error happens in the main process, it will prevent the renderer from displaying a window, so we’ll need a way to clear the previous session without the renderer. Maybe the menu bar will have this option.

## Where to call the migrate function

We will call the migrate function in all the places that we create a redux store and pass in preloaded state. It might be maintainable to make a store enhancer or make our own createStore function that has the migrate function in it.

The places where we create stores:
* In the search page renderer
* In the main process globalStore handler
* In the detail page renderer

@mason-fish I'd love feedback on this here proposal.

**Comment by mason-fish** [comment]
@jameskerr thanks for putting all this together! I'll try and share my initial thoughts in this one comment though they may appear scattered since I can't just comment inline with your proposal.

* The general infrastructure here makes good sense to me. I don't know if we actually need to convert the version to a timestamp though, using the existing incrementing uint seems like plenty and should remain future proof.
* On the topic of "archiving" state, if we know the new app version will not succeed with the old state and that old state failed to transform/migrate to the new shape, wouldn't we want to wipe everything anyway? I do realize that's what we are trying to avoid with this, so we will want to be sure that our migration scripts are well written, but starting the app fresh is at least better than starting it broken. Especially since our current autoupdater will not allow users to stay on an old version, it will update whether they want to or not. This last is also why migrating-down/back doesn't seem necessary.
*  Will there ever be a time when a migration might end up taking a user-impactingly long time? Such as a series of api calls (esp. if we ever had to had to re-ingest something)? If so, and it doesn't seem like a remote possibility, then we will also want to design some sort of progress indicator to the user. I see this all the time when my IDEs re-index files after updating themselves. Maybe it isn't something we need to worry about (at least for now) though.

**Comment by jameskerr** [comment]
> using the existing incrementing uint seems like plenty and should remain future proof.

Well, in the proposal above, the VERSION constant would go away. We would never have to manually increment the version of the state. Each time we run `bin/gen migration` a timestamp at that moment would be generated and put into the file name. The existence of that migration file would indicate that the version has advanced. Does that make sense? I suppose instead of the timestamp, we could increment an integer counter somewhere. But timestamps seem easier to me since they are always advancing.

> On the topic of "archiving" state, if we know the new app version will not succeed with the old state and that old state failed to transform/migrate to the new shape, wouldn't we want to wipe everything anyway?

I see two types of users of the app:
1) We developers: we'd want to see the migration error so that we can fix it
2) End users: they would would want the app to just start up fresh like you suggest

I like the idea of archiving the old state because it would be nice for debugging why something went wrong, or if the user had something they needed in there, then we messed up a migration. The app may start fresh, but at least we can point them to a file somewhere to get their old data.

> This last is also why migrating-down/back doesn't seem necessary.

Agreed.

> Will there ever be a time when a migration might end up taking a user-impactingly long time? 

Good point. I don't see this in the foreseeable future. But yes if we ever have to re-ingest data in the future, a progress bar makes sense.

Thanks for your feedback, Mason.  I'll split this into a few PRs so that we can get more feedback quickly.

**Comment by philrz** [comment]
I've verified in Brim commit `4edb45f` that I can see this logic firing. Bug #815 attributable to migrations was found along the way and is now being pursued.

To test, I started by running Brim GA tag `v0.9.1`, imported a small pcap, and ran a couple queries to populate my History. I closed the app and saved a copy of my `appState.json` and piped it through `jq` for readability. I then upgraded to a Brim release artifact based on commit `4edb45f`. I went into the app and performed another query, then quit and saved a copy of the updated `appState.json`. Comparing the two, I can see the migration made the changes described by @jameskerr in https://github.com/brimsec/brim/pull/793.

```
$ egrep timeFormat\|zeekRunner appState.json-v0.9.1 
[no output]

$ egrep timeFormat\|zeekRunner appState.json-4edb45f 
            "timeFormat": "",
            "zeekRunner": ""
        "zeekRunner": "",
        "timeFormat": ""

$ grep space appState.json-v0.9.1 
              "space": "hello.pcapng.brim",
              "space": "hello.pcapng.brim",
              "space": "hello.pcapng.brim",
        "spaces": {
                    "space": "hello.pcapng.brim",
                    "space": "hello.pcapng.brim",
                    "space": "hello.pcapng.brim",
                "space": "hello.pcapng.brim",
          "space": "hello.pcapng.brim",
          "space": "hello.pcapng.brim",
          "space": "hello.pcapng.brim",
    "spaces": {

$ grep space appState.json-4edb45f 
                "space": "hello.pcapng.brim",
                "space": "hello.pcapng.brim",
                "space": "hello.pcapng.brim",
                "spaceId": "hello.pcapng.brim",
                "spaceName": "hello.pcapng.brim",
          "spaces": {
                      "space": "hello.pcapng.brim",
                      "space": "hello.pcapng.brim",
                      "space": "hello.pcapng.brim",
                      "spaceId": "hello.pcapng.brim",
                      "spaceName": "hello.pcapng.brim",
                  "spaceId": "hello.pcapng.brim",
                  "spaceName": "hello.pcapng.brim",
            "space": "hello.pcapng.brim",
            "space": "hello.pcapng.brim",
            "space": "hello.pcapng.brim",
            "spaceId": "hello.pcapng.brim",
            "spaceName": "hello.pcapng.brim",
      "spaces": {

$ egrep packet_support\|pcap_support appState.json-v0.9.1 
              "packet_support": true,
          "packet_support": true,

$ egrep packet_support\|pcap_support appState.json-4edb45f 
                "pcap_support": true,
            "pcap_support": true,
```

[appState.json-v0.9.1.zip](https://github.com/brimsec/brim/files/4684502/appState.json-v0.9.1.zip)
[appState.json-4edb45f.zip](https://github.com/brimsec/brim/files/4684505/appState.json-4edb45f.zip)


#815 represents one bug we know of related to migrations that still needs to be addressed before we can release.

Thanks @jameskerr!
